Updated build process

jonrandahl opened the pull request:
Resolved incorrect traps for missing env vars in `entrypoint.sh`, added updated commands for `Dockerfile`, as well as added new `SHORTNAME` variable to `Makefile`.

jonrandahl committed:
feat: Updated accessibility statement

Updated accessibility statement to reflect the adjusted release dates for both the expected Qonsole update and preparation timestamp alongside the removal of the deadlines for test revisions. Also includes minor copy changes to resolve typos/punctuation issues.

jonrandahl committed:
style: Updated Welsh translation to improve punctuation

While waiting for the approved translation for the latest update I've updated the copy to improve various characters to be safer to use online.

jonrandahl committed:
docs: Updated changelog and version cadence

jonrandahl committed:
build: Updated Gemfile.lock

jonrandahl committed:
feat: Updated Welsh Translation

As provided by the client's translations as requested

jonrandahl committed:
docs: Updated Changelog 

Updated to reflect date of latest changes

jonrandahl committed:
style: Updated special characters to use entity codes

Refactored the single quotes, apostrophes, and em-dashes to use the appropriate html entity codes to ensure the presentation of the character was maintained due to differences in language use

jonrandahl committed:
feat: Refactored code snippets

Wrapped the appropriate html examples within both a `<code>` wrapper for screen-readability as well as removed erroneous back tick characters

jonrandahl committed:
feat: Removed duplicated main element

Refactored to contain information inside an `<article>` element instead of duplicating the `<main>` element and ID from the template

jonrandahl committed:
Merge pull request #90 from epimorphics/issue/106-hmlr-accessibility-statement-update

HMLR Accessibility Statement Updates

jonrandahl committed:
build: Updated build process

Resolved incorrect traps for missing env vars in `entrypoint.sh`, added updated commands for `Dockerfile`, as well as added new `SHORTNAME` variable to `Makefile`.

jonrandahl committed:
refactor: Removed erroneous comments

jonrandahl committed:
docs: Updated CHANGELOG, README, and version cadence

jonrandahl committed:
build: removed lint and test from make image

jonrandahl committed:
build: cleanup of build files for LR apps

Implements unification approach to Make, Docker and other associated build files to ensure the process completes for specific environments as expected

jonrandahl committed:
build: unified environment configurations across all apps

Updated the matching configuration settings on each environment configuration file to uniify the settings for all apps

jonrandahl committed:
fix: restored merge conflict overwrite

re-added `config.welsh_language_enabled = true`, the feature flag for showing the Welsh language switch affordance

jonrandahl committed:
docs: README update

minor addition of live links for the different apps alongside primarily formatting changes for layout

andrew-pickin-epi commented on entrypoint.sh:
```diff
@@ -28,8 +24,8 @@ then
 
 export RAILS_RELATIVE_URL_ROOT=${APPLICATION_ROOT:-'/app/root'}
-export SCRIPT_NAME=${APPLICATION_ROOT}
+export SCRIPT_NAME=${RAILS_RELATIVE_URL_ROOT}
```
What is the point of setting `SCRIPT_NAME` via this environment variable?
This can only break the system.

jonrandahl replied:
@andrew-pickin-epi `SCRIPT_NAME` is used to "[... ] have both the assets paths and path helpers configured for serving a Rails application from a subdirectory." [^1]

It is similar to `RAILS_RELATIVE_URL_ROOT` in that "[...] `SCRIPT_NAME` isn't available during the asset compilation task so config.relative_url_root needs to be set so that it knows where they will be served from by the webserver. This is especially important for Sass since the asset helpers like image-url, etc. will generate absolute urls." [^2]

I don't follow how this can break the system though and we went through all this together?

[^1]:https://jlintusaari.net/deploying-rails-application-to-subdirectory/
[^2]:https://github.com/rails/rails/issues/8119#issuecomment-25442568

andrew-pickin-epi replied:
You can break the system is `RAILS_RELATIVE_URL_ROOT` different at run-time from what it's set to at build-time (or if either of these is different from `SCRIPT_NAME`). And this is still possible,

andrew-pickin-epi commented on entrypoint.sh:
```diff
@@ -5,20 +5,16 @@ set -e
-then
-  exit 1
-fi
+echo "{\"ts\": $(date -u +%FT%T.%3NZ), \"level\": \"INFO\", \"message\": \"Initiating ${PUBLIC_NAME} application using APPLICATION_ROOT=${APPLICATION_ROOT}, API_SERVICE_URL=${API_SERVICE_URL}}"
```
`APPLICATION_ROOT` is stale here.
API_SERVICE is redundant here.

jonrandahl replied:
This echo was created as a verification flag for the build process and, while I agree and will remove the `APPLICATION_ROOT` variable, having the `API_SERVICE_URL` is good to know at this point in case you're needing to validate any local configuration being passed in at invocation.

jonrandahl replied:
@andrew-pickin-epi Upon further reflection I've not removed the `APPLICATION_ROOT` from the application as it does allow us to pass in a temporary value, like `/`, when trying to get things running locally. 

I would like to say I think this makes the app more robust when developing locally even as an image as that will also match the [development.rb](https://github.com/epimorphics/lr-landing/blob/45608eea14eb2d60b4491f70715000dde287c511/config/environments/development.rb#L49) configuration settings:

```sh
  # The RAILS_RELATIVE_URL_ROOT should be set in the entrypoint.sh script via the 
  # build-time environment variable `APPLICATION_ROOT`; however, if this is not set
  # the following fallback is passed in as a standard root value when the app is run
  # directly via `rails server`
  config.relative_url_root = ENV.fetch('RAILS_RELATIVE_URL_ROOT', '/')
```

Therefore the echo above will allow us to verify the expected value is being used when creating and/or running the image.

N.B. Please forgive me if I've gotten my `build-time` v. `run-time` variables mixed up nonetheless! 🙏

andrew-pickin-epi replied:
It doesn't make it more robust.
If it's possible to set the variable differnetly at build time and run time the it can break,
IF you want to have two setups to run with the without a path then define different environment files. 
Further this IS a function of rails or at least rake. So No APPLICAITON_ROOT makes no sense at all.

jonrandahl replied:
@andrew-pickin-epi while trying to resolve the proxy configuration mentioned previously via further investigation of the older versions of this application, and therefore also the other LR app configurations, we seem to have gotten things slightly muddled up in our approach. 

While I appreciate your guidance throughout these updates, and a lot of good choices were made, I would like to revert the `RAILS_RELATIVE_URL_ROOT` run-time variables to it's [previous configuration](https://github.com/epimorphics/lr-landing/blob/20cac1c7e4f16eeaa27186461faac26081b5bfaf/config/environments/production.rb#L88) based on my findings; let me explain.

Back in [January](https://github.com/epimorphics/lr-landing/commit/20cac1c7e4f16eeaa27186461faac26081b5bfaf) we discussed and agreed that the naming of the variables being passed around during the build process, specifically `RAILS_RELATIVE_URL_ROOT`, was misleading in its use in the build files. We decided then to rename that specific variable to ~~`APPLICATION_PATH`~~ `APPLICATION_ROOT` to be more clear on its purpose.

I, in my Rails naivety, misunderstood the use for that specific name and changed all instances of the RAILS/Rake variable used across the application as well. 

@der kindly highlighted my error to which I revised the renamed variable to only be the ones used in the build files, to which you then updated the validation methods you felt were appropriate in the `entrypoint.sh` file.

Upon further investigation, and in order to improve the quality of the codebase for the new infrastructure, we revised those validation flags together whilst also realising that the run-time value would work if the variable was "locked" to the expected `/app/root` path value in the rails production configuration file. We then cleaned up those validations to only check for the `API_SERVICE_URL` where applicable (funny enough, not in this application).

The Makefile configuration was also refactored as we realised the value being passed in during the precompile phase was being overwritten by the run-time locked value and therefore unnecessary.

While investigating the original configuration of the build files [[52](https://github.com/epimorphics/hmlr-linked-data/issues/52), [72](https://github.com/epimorphics/lr-landing/pull/72/files), [79](https://github.com/epimorphics/hmlr-linked-data/issues/79)] to sort out the proxy I've come to the conclusion that `APPLICATION_ROOT`, and subsequently `RAILS_RELATIVE_URL_ROOT`, should be available to the build process on ALL environments because if you're trying to run the docker image locally without having the proxy running you cannot see the styles, scripting, etc. if you can't change the `RAILS_RELATIVE_URL_ROOT` value; and if you try to run the rails server outside of docker in production mode to use the locked value, it won't build because the `RAILS_SECRET` isn't available.

Could we please discuss this together to agree on an approach that will allow for the most application resilience while still allowing for all of the environments to be run, tested, and validated both inside and out of docker images?

PS. I'm not averse to introducing specific environment files as suggested if that's the best way forward?

andrew-pickin-epi commented on entrypoint.sh:
```diff
@@ -28,8 +24,8 @@ then
+export SCRIPT_NAME=${RAILS_RELATIVE_URL_ROOT}
 
-echo '{"ts": "'"$(date -u +%FT%T.%3NZ)"'", "level": "INFO", "message": "Starting LR Landing application with RAILS_ENV='"${RAILS_ENV}"'", "RAILS_RELATIVE_URL_ROOT"="'"${RAILS_RELATIVE_URL_ROOT}"'", "SCRIPT_NAME"="'"${SCRIPT_NAME}"'", "API_SERVICE_URL"="'"${API_SERVICE_URL}"'"}'
+echo "{\"ts\": $(date -u +%FT%T.%3NZ), \"level\": \"INFO\", \"message\": \"Starting ${PUBLIC_NAME} application with RAILS_ENV=\"${RAILS_ENV}\", \"RAILS_RELATIVE_URL_ROOT\"=\"${RAILS_RELATIVE_URL_ROOT}\", \"SCRIPT_NAME\"=\"${SCRIPT_NAME}\"}"
```
What is `PUBLIC_NAME`?
`RAILS_RELATIVE_URL_ROOT` and `SCRIPT_NAME` should not be used.

der replied:
`PUBLIC_NAME` is set on line 9 (to _LR Landing_) presumably to give one place to change a bit of text that's embedded in two locations.

Not understanding the objection to logging the value of `RAILS_RELATIVE_URL_ROOT` to aid debugging but agreed that `SCRIPT_NAME` is redundant given how it's being set.

jonrandahl replied:
I added `PUBLIC_NAME` to allow better code portability and is similar to `SHORTNAME` in the Makefile; which is used to flag which app is being initiated. 

Again, the echo itself was another verification flag for the build process and should be left as is.

~~I could wrap this in a check for `RAILS_ENV` being `development` if that helps?~~

andrew-pickin-epi commented on README.md:
```diff
@@ -143,15 +147,11 @@ contained changes as well as the ability to diff agains the previous version.
-We use a number of environment variables to determine the runtime behaviour of
-the application:
+We can use a number of environment variables to determine the runtime behaviour
+of the application while developing the codebase locally:
```
Env var table needs re-write.

jonrandahl replied:
I'm working on unpicking the comments and threads found on https://github.com/epimorphics/hmlr-linked-data/issues/52 in order to resolve https://github.com/epimorphics/hmlr-linked-data/issues/79 and your comment.

andrew-pickin-epi commented on README.md:
```diff
@@ -62,20 +63,21 @@ This will run through the assets precompilation, linting and testing before
 With the proxy and Docker container running you can access the application as
-[`localhost:3001/app/root/`](http://localhost:3001/app/root/)
-(note the trailing path).
+[`localhost:3001/app/root/`](http://localhost:3001/app/root/) (note the trailing
```
This is inconsistent with ports in production.
Move proxy from 3001.

jonrandahl replied:
I agree, I used `localhost:3030` in my changes which I will push up in a new PR for the simple web proxy repository and update accordingly where needed.

andrew-pickin-epi replied:
I've reconsidered this. I think the proxy should listen on 8080.

jonrandahl replied:
The issue with using `8080` is that's the API port: `http://localhost:8080` set in the `API_SERVICE_URL` default value? 

Mind you I only went with `3030` as it was still in the 3k range but we could choose something that works for all the different endpoints?

der reviewed (approved):
Can't really comment the `SCRIPT_NAME` discussion. Have attempted to understand where that might be used without success. If it's not actually needed then better to get rid of it but if testing shows it is need then so be it.

Once that's resolved then happy for this to merge.

jonrandahl committed:
docs: README update

Reorganised and adjusted content to be clearer for local development

jonrandahl committed:
Merge branch 'dev' into spike/resolve-failed-build

Brings the latest accessibility changes from the dev branch into the current updates

jonrandahl committed:
docs: Updated CHANGELOG

Brings the latest CHANGELOG changes from the dev branch in the correct order of release.

jonrandahl committed:
build: Clarified use of Run-time env var purpose

jonrandahl committed:
Merge remote-tracking branch 'origin/dev-infrastructure' into spike/resolve-failed-build

jonrandahl committed:
build: resolving `config.relative_url_root` 

Updated to handle locking the root path for individual environments as well as removing unnecessary environment variables

jonrandahl committed:
build: better logging on start-up

jonrandahl committed:
docs: Updated CHANGELOG and README

Rewritten to reflect the latest approach for both local and dockerised invocations of the application. Recorded summaries of changes for future reference.

jonrandahl committed:
docs: Fixed typo

jonrandahl committed:
Merge remote-tracking branch 'origin/dev-infrastructure' into spike/resolve-failed-build

jonrandahl committed:
Merge branch 'dev-infrastructure' into spike/resolve-failed-build

jonrandahl committed:
Merge branch 'dev-infrastructure' into spike/resolve-failed-build

jonrandahl committed:
refactor: updated robocop rules

jonrandahl committed:
fix: adjusted error reporting environments

jonrandahl committed:
build: environment  specific configurations

jonrandahl committed:
build: streamlined entrypoint script 

also includes adjusted comment on Dockerfile

jonrandahl committed:
build: Updated gems and lock file

Specific to Epimorphics internal gems

jonrandahl committed:
docs: Improved README

andrew-pickin-epi commented on entrypoint.sh:
```diff
@@ -8,11 +8,14 @@ mkdir -pm 1777 ./tmp
+[ "$RAILS_ENV" == "production" ] && [ -z "$SECRET_KEY_BASE" ] && SECRET_KEY_BASE=$(./bin/rails secret) && export SECRET_KEY_BASE
 
-[ -n "${RAILS_RELATIVE_URL_ROOT}" ] && echo "{\"ts\":\"$(date -u +%FT%T.%3NZ)\",\"level\":\"INFO\",\"message\":\"RAILS_RELATIVE_URL_ROOT=${RAILS_RELATIVE_URL_ROOT}\"}"
+[ -n "$RAILS_RELATIVE_URL_ROOT" ] && echo "{\"ts\":\"$(date -u +%FT%T.%3NZ)\",\"level\":\"INFO\",\"message\":\"RAILS_RELATIVE_URL_ROOT=${RAILS_RELATIVE_URL_ROOT}\"}"
```
Does landing work with RAIL_RELATIVE_URL_ROOT work with anything but '/'.
Is this machinery and log entry relevant for Landing?

jonrandahl replied:
@andrew-pickin-epi this was felt to be logical and safe *if* a RAILS_RELATIVE_URL_ROOT value was passed in we would report the _non-default_ value

jonrandahl committed:
build: cleaned BUNDLE_CONFIG

Removed `${HOME}/` from path

andrew-pickin-epi commented on README.md:
````diff
@@ -2,160 +2,186 @@
 
 ```sh
-rails server -e production -p <port> -b 0.0.0.0
+API_SERVICE_URL=http://localhost:8888 rails server
````
Why are you referencing API_SERVICE_URL in this repository when it's not used here. 
Further the, where it is being used we should attempt to use the internal docker name as reference in the Makefile, unless there is a reason to used something else when attempting to set up some alternative routing, in which case the purpose of the alternative should be made clear.

jonrandahl replied:
Andy, when you're running locally in Rails and you want to test Qonsole then you need to pass in the API_SERVICE_URL or else Qonsole doesn't have the api access needed as per the comment above states.

andrew-pickin-epi commented on README.md:
```diff
@@ -2,160 +2,186 @@
 - [Qonsole (Rails)](https://github.com/epimorphics/qonsole-rails)
 
-## Production Mode
+## Developer notes
```
There is a section here call ed "Developer notes" but the contents are not just for a audience of developers, this is a bad choice of title

jonrandahl committed:
docs: Updated README header
